### ISSUE TITLE
Build correct SQLAlchemy URI in hooks based on `DbApiHook`

### ISSUE THREAD

**Issue by Taragolis** [apache/airflow#38195]
### Body

Original discussion: https://lists.apache.org/thread/8rhmz3qh30hvkondct4sfmgk4vd07mn5

tl;dr; `DbApiHook` assumes that Airflow Connection URI representation is a valid SA URI, but it is not true and never was, it might work in simple cases, however it will fail in some complex one.

https://github.com/apache/airflow/blob/2ab60812a040cb8b760acaf396f625d0d719e4be/airflow/providers/common/sql/hooks/sql.py#L184-L192

This task intend to track fixes into the hooks which based on `DbApiHook`.

Hooks which don't overwrite invalid `DbApiHook.get_uri` implementation
---

- [x] `MySqlHook` from `mysql` provider. This hook provide support for two different drivers.
   + [ ] `mysqlclient`: https://pypi.org/project/mysqlclient/
   + [ ] `mysql-connector-python`: https://pypi.org/project/mysql-connector-python/
- [x] OracleHook from `oracle` provider
- [x] JdbcHook from `jdbc` provider
- [x] TrinoHook from `trino` provider
- [ ] HiveServer2Hook from `apache.hive` provider
- [ ] ImpalaHook from `apache.impala` provider
- [ ] DatabricksSqlHook from `databricks` provider
- [ ] ExasolHook from `exasol` provider
- [ ] PrestoHook from `presto` provider
- [ ] VerticaHook from `vertica` provider
- [ ] TeradataHook from `teradata` provider

Hooks which uses `Connection.get_uri` for construct SQLAlchemy URI
---

- [ ] PostgresHook from `postgres` provider. https://github.com/apache/airflow/issues/38187, https://github.com/apache/airflow/issues/34456
- [ ] MsSqlHook from `microsoft.mssql` provider
- [ ] SqliteHook from `sqlite` provider

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

**Comment by Bowrna** [comment]
hello @Taragolis I was checking this issue and seeing if I could fix it.

https://github.com/apache/airflow/blob/0e8f108313d4af0b450581661aeb8ed35e82a8e6/airflow/models/connection.py#L271-L329

Is fixing the get_uri part so that  Airflow Connection URI representation becomes a valid SA URI the right way to fix this issue?

**Comment by rawwar** [comment]
@Taragolis , I am working on this PR - https://github.com/apache/airflow/pull/38871 . This is a separate PR to just update the `DbApiHook` to include a method to return a SqlAlchemy URL that can be reused.

**Comment by dabla** [comment]
> As far as i remember, we recently discuss this case in slack. I guess better introduce `sa_uri` property (or some similar naming) which returns [`sqlalchemy.engine.URL`](https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.engine.URL) into the `DbApiHook` without any implementation, e.g. raise `NotImplementedError` and implements it per hook which have implementation for SQAlchemy.
> 
> `get_uri` - not clear which URI it is implements, DBApi? SQLAlchemy or even some internal ([ElasticsearchSQLHook](https://github.com/apache/airflow/blob/dec2662190dd4480d0c631da733e19d2ec9a479d/airflow/providers/elasticsearch/hooks/elasticsearch.py#L113))
> 
> cc: @potiuk @dstandish

Why not implement another method named get_url and use that one to instantiate the SQLALchemy engine, as the engine can accept an SQLAlchemy URL object.  Also the SQLAlchemy URL class has a factory method named create, which easily allows you to create the URL without the hastle of string manipulations, this is how I did it for my custom SQL operator (the one I mentioned on Slack):

```
    def get_url(self) -> URL:
        return URL.create(
            self.scheme,
            username=self.connection.login,
            password=self.connection.password,
            host=self.connection.host,
            port=self.connection.port,
        )
```

Also see [here ](https://docs.sqlalchemy.org/en/20/core/engines.html#creating-urls-programmatically)the SQL documentation concerning creating URL's programmatically: [https://docs.sqlalchemy.org/en/20/core/engines.html#creating-urls-programmatically](https://docs.sqlalchemy.org/en/20/core/engines.html#creating-urls-programmatically)

**Comment by Taragolis** [comment]
There is new property which intends to returns SQL Alchemy URL, current behaviour of `get_url` is ambiguous and non-all hook return SA URI here, for some of the DB there is no even SA dialects exists, so better keep `get_url` as is and step-by-step migrate to appropriate ones. So if dialect exists and it installed with provider I do not have any objections to add it implementation as `sqlalchemy_url`

**Comment by dabla** [comment]
> There is new property which intends to returns SQL Alchemy URL, current behaviour of `get_url` is ambiguous and non-all hook return SA URI here, for some of the DB there is no even SA dialects exists, so better keep `get_url` as is and step-by-step migrate to appropriate ones. So if dialect exists and it installed with provider I do not have any objections to add it implementation as `sqlalchemy_url`

Hey Andrey think there is so confusion here.  Yes we should definitly leave the get_uri method as it is (not get_url as it doesn’t exist yet) and implement a new get_url method which would then be called by the existing get_sqlalchemy_engine.

**Comment by dabla** [comment]
This is how I done it in my custom SQLRowsInsertOperator:

```
    @cached_property
    def connection(self) -> Connection:
        return BaseHook.get_connection(self.conn_id)  # type: ignore

    @cached_property
    def scheme(self) -> str:
        return self.connection.extra_dejson["sqlalchemy_scheme"]

    @cached_property
    def table_name(self) -> str:
        return self.get_columns().table_name_with_schema()

    @cached_property
    def driver(self) -> str:
        return self.connection.extra_dejson["driver"]

    # TODO: Maybe this should be moved to DbAPiHook
    def get_url(self) -> URL:
        self.log.info("Connection schema: %s", self.schema)
        self.log.info("Connection scheme: %s", self.scheme)
        self.log.info("Connection driver: %s", self.driver)
        self.log.info("Connection type: %s", self.connection.conn_type)
        self.log.info("Connection login: %s", self.connection.login)
        self.log.info("Connection password: %s", self.connection.password)
        self.log.info("Connection host: %s", self.connection.host)
        self.log.info("Connection port: %s", self.connection.port)

        return URL.create(
            self.scheme,
            username=self.connection.login,
            password=self.connection.password,
            host=self.connection.host,
            port=self.connection.port,
        )

   # TODO: Maybe this method in DbAPiHook should be changed to this instead of calling get_uri() method
    def get_sqlalchemy_engine(self, engine_kwargs=None):
        if engine_kwargs is None:
            engine_kwargs = {}
        return create_engine(self.get_url(), **engine_kwargs)

    # TODO: Maybe this should be moved to DbAPiHook
    @cached_property
    def inspector(self):
        engine = self.get_sqlalchemy_engine()
        self.log.debug("Engine drivername: %s", engine.url.drivername)
        self.log.debug("Engine username: %s", engine.url.username)
        self.log.debug("Engine password: %s", engine.url.password)
        self.log.debug("Engine host: %s", engine.url.host)
        self.log.debug("Engine port: %s", engine.url.port)
        self.log.debug("Engine database: %s", engine.url.database)
        self.log.debug("Engine dialect: %s", engine.dialect.name)

        return Inspector.from_engine(engine)
```

**Comment by rawwar** [comment]
>     # TODO: Maybe this should be moved to DbAPiHook
>     def get_url(self) -> URL:
>         self.log.info("Connection schema: %s", self.schema)
>         self.log.info("Connection scheme: %s", self.scheme)
>         self.log.info("Connection driver: %s", self.driver)
>         self.log.info("Connection type: %s", self.connection.conn_type)
>         self.log.info("Connection login: %s", self.connection.login)
>         self.log.info("Connection password: %s", self.connection.password)
>         self.log.info("Connection host: %s", self.connection.host)
>         self.log.info("Connection port: %s", self.connection.port)
> 
>         return URL.create(
>             self.scheme,
>             username=self.connection.login,
>             password=self.connection.password,
>             host=self.connection.host,
>             port=self.connection.port,
>         )
> 

I think this is the plan. First, to add `sqlalchemy_url` to all the providers that use it. And, then, to modify to the base class if I am not wrong.

**Comment by rawwar** [comment]
> Hey Andrey think there is some confusion here. Yes we should definitly leave the get_uri method as it is (not get_url as it doesn’t exist yet) and implement a new get_url method which would then be called by the existing get_sqlalchemy_engine.

@dabla , I think, we are indeed adding `get_url` functionality using `sqlalchemy_url` property.

**Comment by guan404ming** [comment]
Hi, I think I could help with this issue~ 
Also, I have submitted some PR toward this issues

**Comment by eladkal** [comment]
@guan404ming what are the open items left on this issue? I see you handled several of them

**Comment by guan404ming** [comment]
I plan to get back here after `get_df` migration and I think I would handle the remaining later.

**Comment by BrianLii** [comment]
Hello, I would like to work on this issue.

**Comment by BrianLii** [comment]
@eladkal I can confirm the open items left on this issue:

- [X] MySqlHook from `mysql` provider.
- [X] OracleHook from `oracle` provider
- [X] JdbcHook from `jdbc` provider
- [X] TrinoHook from `trino` provider
- [ ] HiveServer2Hook from `apache.hive` provider
- [ ] ImpalaHook from `apache.impala` provider
- [x] DatabricksSqlHook from `databricks` provider 
- [x] ExasolHook from `exasol` provider
- [ ] PrestoHook from `presto` provider
- [x] VerticaHook from `vertica` provider
- [ ] TeradataHook from `teradata` provider

**Comment by eladkal** [comment]
@sc250072 can you or someone else from Teradata handle the TeradataHook task?

**Comment by BrianLii** [comment]
I don’t see a need to re-implement the `get_uri` feature for `HiveServer2`. According to the [Airflow HiveServer2 connection docs](https://airflow.apache.org/docs/apache-airflow-providers-apache-hive/stable/connections/hiveserver2.html), the connection URI format is compatible with the existing `get_uri` in `DbApiHook`. However, I can implement the currently missing `sqlalchemy_url` property to return the `get_uri` value.

**Comment by eladkal** [comment]
@BrianLii are you working on the remaining items?

**Comment by BrianLii** [comment]
Hi, I'll be working on this issue this week. I had to switch to another task recently, but I'll return to this now. Thank you for the reminder.

**Comment by BrianLii** [comment]
Hi, I may not be able to work on this issue in the near term. Could you please unassign me?

**Comment by Pyasma** [comment]
hi i want to work on this issue could you please assign it to me

**Comment by sc250072** [comment]
> [@sc250072](https://github.com/sc250072) can you or someone else from Teradata handle the TeradataHook task?

@eladkal let us handle it and let you know

**Comment by eladkal** [comment]
@Pyasma will you work on the remaining items as well?

**Comment by Pyasma** [comment]
Yea I am working on writing tests for presto hooks

**Comment by sc250072** [comment]
> > [@sc250072](https://github.com/sc250072) can you or someone else from Teradata handle the TeradataHook task?
> 
> [@eladkal](https://github.com/eladkal) let us handle it and let you know

@eladkal  https://github.com/apache/airflow/pull/56305 to address this for Teradata

**Comment by eladkal** [comment]
Remaining Tasks:

- [x] HiveServer2Hook from `apache.hive` provider https://github.com/apache/airflow/pull/59878
- [x] ImpalaHook from `apache.impala` provider https://github.com/apache/airflow/pull/57153
- [x] PrestoHook from `presto` provider https://github.com/apache/airflow/pull/56298
- [x] TeradataHook from `teradata` provider https://github.com/apache/airflow/pull/56305

**Comment by Pyasma** [comment]
Hey i will start working on the last Hook

**Comment by Pyasma** [comment]
@potiuk can we close this issue now ?

**Comment by Pyasma** [comment]
@eladkal can we close this issue ?

**Comment by potiuk** [comment]
Yep. Closing it now :)
